packages/blocks/src/page-block/edgeless/selection-manager/default.ts

```diff
@@ -0,0 +1,210 @@
+} from '../../utils/cursor';
+import { showFormatQuickBar } from '../../../components/format-quick-bar';
+import { SelectionController } from './index';
+
+export class DefaultSelectionController extends SelectionController<DefaultMouseMode> {
+  readonly mouseMode = <DefaultMouseMode>{
+    type: 'default',
+  };
+
+  private _startRange: Range | null = null;
+
+  private _updateFrameSelectionState(x: number, y: number) {
+    if (this._frameSelectionState) {
+      this._frameSelectionState.end = new DOMPoint(x, y);
+    }
+    if (this._hoverState) {
+      this._blockSelectionState = {
+        type: 'single',
+        selected: this._hoverState.block,
+        viewport: this._container.viewport,
+        rect: this._hoverState.rect,
+        active: false,
+      };
+    }
+  }
+
+  private _updateHoverState(hoverBlock: RootBlockModel | null) {
+    if (hoverBlock) {
+      this._hoverState = {
+        rect: getSelectionBoxBound(this._container.viewport, hoverBlock.xywh),
+        block: hoverBlock,
+      };
+    } else {
+      this._hoverState = null;
+    }
+  }
+
+  private _handleClickOnSelected(selected: RootBlockModel, e: SelectionEvent) {
+    const { viewport } = this._container;
+
+    switch (this.blockSelectionState.type) {
+      case 'none':
+        this._blockSelectionState = {
+          type: 'single',
+          active: false,
+          viewport: this._container.viewport,
+          selected,
+          rect: getSelectionBoxBound(viewport, selected.xywh),
+        };
+        this._container.signals.updateSelection.emit(this.blockSelectionState);
+        break;
+      case 'single':
+        if (this.blockSelectionState.selected === selected) {
+          this.blockSelectionState.active = true;
+          this._container.signals.updateSelection.emit(
+            this.blockSelectionState
+          );
+        } else {
+          this._blockSelectionState = {
+            type: 'single',
+            viewport: this._container.viewport,
+            active: false,
+            selected,
+            rect: getSelectionBoxBound(viewport, selected.xywh),
+          };
+          this._container.signals.updateSelection.emit(
+            this.blockSelectionState
+          );
+        }
+        handleNativeRangeClick(this._space, e);
+        break;
+    }
+  }
+
+  onContainerClick(e: SelectionEvent): void {
+    const { viewport } = this._container;
+    const [modelX, modelY] = viewport.toModelCoord(e.x, e.y);
+    const selected = pick(this._blocks, modelX, modelY, this._container, e);
+
+    if (selected) {
+      this._handleClickOnSelected(selected, e);
+    } else {
+      this._blockSelectionState = { type: 'none' };
+      this._container.signals.updateSelection.emit(this.blockSelectionState);
+      resetNativeSelection(null);
+    }
+  }
+
+  onContainerContextMenu(e: SelectionEvent): void {
+    repairContextMenuRange(e);
+  }
+
+  onContainerDblClick(e: SelectionEvent): void {
+    // do nothing
```

Comment:
We express this kind of intention with `noop()` explicitly. This applies to other `// do nothing` lines.